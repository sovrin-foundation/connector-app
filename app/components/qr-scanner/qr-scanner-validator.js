// @flow
import schemaValidator from '../../services/schema-validator'
import {
  QR_CODE_SENDER_DID,
  QR_CODE_SENDER_VERIFICATION_KEY,
  QR_CODE_LOGO_URL,
  QR_CODE_REQUEST_ID,
  QR_CODE_SENDER_KEY_DELEGATION,
  QR_CODE_SENDER_NAME,
  QR_CODE_TARGET_NAME,
  QR_CODE_SENDER_DETAIL,
  QR_CODE_DELEGATION_DID,
  QR_CODE_DELEGATION_KEY,
  QR_CODE_DELEGATION_SIGNATURE,
  QR_CODE_SENDER_AGENCY,
  QR_CODE_SENDER_AGENCY_DID,
  QR_CODE_SENDER_AGENCY_KEY,
  QR_CODE_SENDER_AGENCY_ENDPOINT,
} from '../../api/api-constants'
import type { QrCode } from './type-qr-scanner'

const qrSchema = {
  type: 'object',
  properties: {
    [QR_CODE_REQUEST_ID]: { type: 'string' },
    [QR_CODE_SENDER_DETAIL]: {
      type: 'object',
      properties: {
        [QR_CODE_SENDER_DID]: { type: 'string' },
        [QR_CODE_SENDER_NAME]: { type: 'string' },
        [QR_CODE_LOGO_URL]: { type: 'string' },
        [QR_CODE_SENDER_VERIFICATION_KEY]: { type: 'string' },
        [QR_CODE_SENDER_KEY_DELEGATION]: {
          type: 'object',
          properties: {
            [QR_CODE_DELEGATION_DID]: { type: 'string' },
            [QR_CODE_DELEGATION_KEY]: { type: 'string' },
            [QR_CODE_DELEGATION_SIGNATURE]: { type: 'string' },
          },
          required: [
            QR_CODE_DELEGATION_DID,
            QR_CODE_DELEGATION_KEY,
            QR_CODE_DELEGATION_SIGNATURE,
          ],
        },
      },
      required: [
        QR_CODE_SENDER_DID,
        QR_CODE_SENDER_NAME,
        QR_CODE_LOGO_URL,
        QR_CODE_SENDER_VERIFICATION_KEY,
        QR_CODE_SENDER_KEY_DELEGATION,
      ],
    },
    [QR_CODE_SENDER_AGENCY]: {
      type: 'object',
      properties: {
        [QR_CODE_SENDER_AGENCY_DID]: { type: 'string' },
        [QR_CODE_SENDER_AGENCY_KEY]: { type: 'string' },
        [QR_CODE_SENDER_AGENCY_ENDPOINT]: { type: 'string' },
      },
      required: [
        QR_CODE_SENDER_AGENCY_DID,
        QR_CODE_SENDER_AGENCY_KEY,
        QR_CODE_SENDER_AGENCY_ENDPOINT,
      ],
    },
    [QR_CODE_TARGET_NAME]: { type: 'string' },
  },
  required: [
    QR_CODE_REQUEST_ID,
    QR_CODE_SENDER_DETAIL,
    QR_CODE_SENDER_AGENCY,
    QR_CODE_TARGET_NAME,
  ],
}

// this is the number that is max allowed by 200 X 200 qr code
// if we are getting more than that, then there is an issue from
// qr code generation site or someone is trying to break the app
const qrCodeAllowedLength = 1200

// If we need to use QR scanner somewhere else in app as well
// remove this method and create an equivalent in another
// component as well, for now this is good here
export function isValidQrCode(qr: string): QrCode | boolean {
  // check the length of qr data so that we don't get memory issues
  // while parsing the json
  if (qr.length >= qrCodeAllowedLength) {
    return false
  }

  let qrData: QrCode
  try {
    qrData = (JSON.parse(qr): QrCode)
  } catch (e) {
    // qr code is not a valid json, we still return false
    return false
  }

  // check if qr code schema is valid
  if (!schemaValidator.validate(qrSchema, qrData)) {
    return false
  }

  // since all checks are passed
  // qr code data is valid, send the parsed data back
  return qrData
}
