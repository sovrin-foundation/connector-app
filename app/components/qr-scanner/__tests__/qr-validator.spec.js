// @flow
import { isValidQrCode } from '../qr-scanner-validator'

describe('QR Scanner validator', () => {
  it('should return false if schema is not valid at top level', () => {
    const invalidQrSchemas = [
      { n: '' },
      { p: '' },
      { n: '', a: '' },
      { n: '', a: '' },
      { p: '', '1': '' },
      // invalid as per length
      {
        n: '',
        p: 'a'.repeat(1300),
      },
      // invalid at second level
      { p: JSON.stringify({ lu: '' }), n: '', sD: '', sVk: '' },
      { p: JSON.stringify({ lu: '', rid: '' }), n: '', sD: '', sVk: '' },
      {
        p: JSON.stringify({ lu: '', rid: '', sakdp: '' }),
        n: '',
        sD: '',
        sVk: '',
      },
      {
        p: JSON.stringify({ lu: '', rid: '', sakdp: '', sn: '' }),
        n: '',
        sD: '',
        sVk: '',
      },
      {
        p: JSON.stringify({ lu: '', rid: '', sakdp: '', sn: '', ts: '' }),
        n: '',
        sD: '',
        sVk: '',
      },
    ]

    invalidQrSchemas.map(schema =>
      expect(isValidQrCode(JSON.stringify(schema))).toBe(false)
    )
  })

  it('should return data for valid schema', () => {
    const validSchema = {
      id: 'yta2odh',
      s: {
        n: 'ent-name',
        dp: {
          d: 'N2Uyi6SVsHZq1VWXuA3EMg',
          k: 'CTfF2sZ5q4oPcBvTP75pgx3WGzYiLSTwHGg9zUsJJegi',
          s:
            '/FxHMzX8JaH461k1SI5PfyxF5KwBAe6VlaYBNLI2aSZU3APsiWBfvSC+mxBYJ/zAhX9IUeTEX67fj+FCXZZ2Cg==',
        },
        d: 'F2axeahCaZfbUYUcKefc3j',
        l: 'ent-logo-url',
        v: '74xeXSEac5QTWzQmh84JqzjuXc8yvXLzWKeiqyUnYokx',
      },
      sa: {
        d: 'BDSmVkzxRYGE4HKyMKxd1H',
        v: '6yUatReYWNSUfEtC2ABgRXmmLaxCyQqsjLwv2BomxsxD',
        e: '52.38.32.107:80/agency/msg',
      },
      t: 'there',
    }

    expect(isValidQrCode(JSON.stringify(validSchema))).toMatchSnapshot()
  })
})
