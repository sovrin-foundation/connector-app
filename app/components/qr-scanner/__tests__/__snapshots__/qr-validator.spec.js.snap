// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QR Scanner validator should return data for valid schema 1`] = `
Object {
  "id": "yta2odh",
  "s": Object {
    "d": "F2axeahCaZfbUYUcKefc3j",
    "dp": Object {
      "d": "N2Uyi6SVsHZq1VWXuA3EMg",
      "k": "CTfF2sZ5q4oPcBvTP75pgx3WGzYiLSTwHGg9zUsJJegi",
      "s": "/FxHMzX8JaH461k1SI5PfyxF5KwBAe6VlaYBNLI2aSZU3APsiWBfvSC+mxBYJ/zAhX9IUeTEX67fj+FCXZZ2Cg==",
    },
    "l": "ent-logo-url",
    "n": "ent-name",
    "v": "74xeXSEac5QTWzQmh84JqzjuXc8yvXLzWKeiqyUnYokx",
  },
  "sa": Object {
    "d": "BDSmVkzxRYGE4HKyMKxd1H",
    "e": "52.38.32.107:80/agency/msg",
    "v": "6yUatReYWNSUfEtC2ABgRXmmLaxCyQqsjLwv2BomxsxD",
  },
  "t": "there",
}
`;
